### PR TITLE
Improve zooming with the mouse wheel

### PR DESCRIPTION
**Pull request by calixteman**
- Refactor the text layer code in order to avoid compute it on each draw: the idea is just to resuse what we got on the first draw. Now, we only update the scaleX of the different spans and the other values are dependant of --scale-factor.
- Move some properties in the CSS in order to avoid any updates in JS.
- When zooming in using mouse wheel, delay the page drawing and rescale in using css in order to minimize the drawing operations.
- with tracemonkey.pdf, we can see a real difference when profiling zooming operations (in using ctrl+wheel): there is nothing in the worker and this patch helps to remove all the jank markers.

### CODE REVIEW

File: web/pdf_viewer.js [hunk 1]
```diff
@@ -1020,7 +1029,7 @@ class PDFViewer {
     this.update();
   }
 
-  #scrollIntoView(pageView, pageSpot = null) {
+  #scrollIntoView(pageView, timeout = null, pageSpot = null) {
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1032]: All of these scroll-related changes makes me a fair bit worried, since delaying scrolling feels like something that could *potentially* cause other unforeseen issues.
Note that much of this code is quite old and there's lots of edge-cases to consider, so lots more comments explaining the logic would seem like a necessity here.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Please split this into a few different PRs, or at the very least a few different commits, since there's just way too many things changing all at once which feels *very* risky and thus really difficult to review.

*Edit:* In particular, implementing #15131 *separately* first would probably improve things a fair bit already.

File: src/display/annotation_layer.js [hunk 2]
```diff
@@ -2659,29 +2658,12 @@ class AnnotationLayer {
...
-    this.#setDimensions(div, viewport);
     this.#setAnnotationCanvasMap(div, annotationCanvasMap);
     div.hidden = false;
   }
 
-  /**
-   * @param {HTMLDivElement} div
-   * @param {PageViewport} viewport
-   */
-  static #setDimensions(div, { width, height, rotation }) {
```
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:2673]: Here, and elsewhere, why are these methods being moved from the `src/`-folder and into the `web/`-folder? Won't that make the core PDF.js library less usable, assuming that we still care about non-Firefox use-cases here?

Comment (calixteman) [reply src/display/annotation_layer.js:2673]: I changed the dimensions to something using the page dimensions and `--scale-factor` so now they're constant whatever we do (zooming, rotating, ...). So I thought it'd make more sense to have this code where the div properties are initialized.

File: web/pdf_viewer.js (hunk 1 above)
Comment (calixteman) [reply web/pdf_viewer.js]: Ok, I'll see later how to improve that stuff.

File: src/display/text_layer.js [hunk 3]
```diff
@@ -46,23 +47,43 @@ const DEFAULT_FONT_SIZE = 30;
 const DEFAULT_FONT_ASCENT = 0.8;
 const ascentCache = new Map();
 
-function getAscent(fontFamily, ctx) {
+function getCtx(size, isOffscreenCanvasSupported, alpha = true) {
+  let ctx;
+  if (isOffscreenCanvasSupported && FeatureTest.isOffscreenCanvasSupported) {
+    ctx = new OffscreenCanvas(size, size).getContext("2d", { alpha });
```
Comment (Snuffleupagus) [inline src/display/text_layer.js:53]: Do we "need" to use `isOffscreenCanvasSupported` here, i.e. are the any clear benefits?
If not, maybe do that in a separate commit to reduce the size of these changes such that reviewing becomes slightly easier?

File: web/pdf_viewer.css [hunk 4]
```diff
@@ -141,6 +141,10 @@
   display: none;
 }
 
+.pdfViewer .page .textLayer[hidden] {
+  display: none;
+}
+
```
Comment (Snuffleupagus) [inline web/pdf_viewer.css:154]: All other layers simply do `div.hidden = true` in their respective `web/` JS-files, so can we please use the same pattern for this as well for consistency?

**Review by Snuffleupagus — COMMENTED** [review]
As mentioned in the edit in https://github.com/mozilla/pdf.js/pull/15717#pullrequestreview-1188518849, can we please move the first commit (i.e. for addressing issue #15131) into its own PR?

File: src/display/text_layer.js (hunk 3 above)
Comment (calixteman) [reply src/display/text_layer.js:53]: Please move your question here: https://github.com/mozilla/pdf.js/pull/15722 and I'll answer.

File: web/pdf_viewer.css (hunk 4 above)
Comment (calixteman) [reply web/pdf_viewer.css:154]: Same as above.
